test/DateYearDurationTest.cpp

```diff
@@ -0,0 +1,977 @@
+  DateYearOrDuration d5{
+      DayTimeDuration{DayTimeDuration::Type::Positive, 1, 24, 23, 62.44}};
+  DateYearOrDuration d6{
+      DayTimeDuration{DayTimeDuration::Type::Negative, 1, 24, 23, 62.44}};
+  DateYearOrDuration d7{
+      DayTimeDuration{DayTimeDuration::Type::Negative, 1, 25, 23, 62.44}};
+
+  ASSERT_EQ(d1, d1);
+  ASSERT_EQ(d2, d2);
+  ASSERT_EQ(d3, d3);
+  ASSERT_EQ(d4, d4);
+  ASSERT_EQ(d5, d5);
+  ASSERT_EQ(d6, d6);
+  ASSERT_EQ(d7, d7);
+  ASSERT_LT(d1, d2);
+  ASSERT_LT(d2, d3);
+  ASSERT_LT(d1, d3);
+  ASSERT_LT(d4, d5);
+  ASSERT_LT(d6, d5);
+  ASSERT_LT(d7, d4);
+  ASSERT_LT(d7, d6);
+}
+
+// SECTION TEST DURATION
+//______________________________________________________________________________
+TEST(DayTimeDuration, sizeInitDayTimeDuration) {
+  ASSERT_EQ(sizeof(DayTimeDuration), 8);
+  ASSERT_EQ(DayTimeDuration::numUnusedBits, 16);
+  DayTimeDuration defaultDuration{};
+  ASSERT_TRUE(defaultDuration.isPositive());
+  EXPECT_NEAR(defaultDuration.getSeconds(), 0.00, 0.001);
+  ASSERT_EQ(defaultDuration.getMinutes(), 0);
+  ASSERT_EQ(defaultDuration.getHours(), 0);
+  ASSERT_EQ(defaultDuration.getDays(), 0);
+}
+
+//______________________________________________________________________________
+TEST(DayTimeDuration, setAndGetValues) {
+  ad_utility::SlowRandomIntGenerator randomDay{0, 1048575};
+  ad_utility::SlowRandomIntGenerator randomHour{0, 23};
+  ad_utility::SlowRandomIntGenerator randomMinute{0, 59};
+  ad_utility::RandomDoubleGenerator randomSecond{0, 59.9990};
+  DayTimeDuration::Type positive = DayTimeDuration::Type::Positive;
+  DayTimeDuration::Type negative = DayTimeDuration::Type::Negative;
+  size_t i = 0, max = 3333;
+  while (i < max) {
+    auto seconds = randomSecond();
+    auto minutes = randomMinute();
+    auto days = randomDay();
+    auto hours = randomHour();
+    DayTimeDuration duration1{positive, days, hours, minutes, seconds};
+    DayTimeDuration duration2{negative, days, hours, minutes, seconds};
+    EXPECT_NEAR(duration1.getSeconds(), seconds, 0.001);
+    EXPECT_NEAR(duration2.getSeconds(), seconds, 0.001);
+    ASSERT_TRUE(duration1.isPositive());
+    ASSERT_FALSE(duration2.isPositive());
+    ASSERT_EQ(duration1.getDays(), days);
+    ASSERT_EQ(duration2.getDays(), days);
+    ASSERT_EQ(duration1.getHours(), hours);
+    ASSERT_EQ(duration2.getHours(), hours);
+    ASSERT_EQ(duration1.getMinutes(), minutes);
+    ASSERT_EQ(duration2.getMinutes(), minutes);
+
+    // Basic comparison
+    // duration1 is set to positive, duration2 to negative, this should
+    // always hold.
+    ASSERT_TRUE(duration1 == duration1);
+    ASSERT_TRUE(duration2 == duration2);
+    ASSERT_TRUE(duration2 < duration1);
+    ASSERT_FALSE(duration2 == duration1);
+    i++;
+  }
+}
+
+//______________________________________________________________________________
+TEST(DayTimeDuration, checkParseAndGetStringForSpecialValues) {
+  DayTimeDuration duration0 =
```

Comment:
This can be made much more readable if you

1. Implement a lambda (best a gmock matcher) that takes the input, the expected literal, and the expected datatype.

The you can write something like
```
EXPECT_THAT("PT0H0M0S", matchesStringAndType("PT0S", XSD_DAYTIME_DURATION_TYPE).
```

Let me know if you need help with this.